server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java

```diff
@@ -0,0 +1,350 @@
+        final List<CuratorOp> deleteOps = new ArrayList<>(pathsCreatedInThisAnnouncer.size());
+        for (String parent : pathsCreatedInThisAnnouncer) {
+          try {
+            deleteOps.add(curator.transactionOp().delete().forPath(parent));
+          }
+          catch (Exception e) {
+            log.error(e, "Unable to delete parent[%s].", parent);
+          }
+        }
+        try {
+          curator.transaction().forOperations(deleteOps);
+        }
+        catch (Exception e) {
+          log.error(e, "Unable to commit transaction.");
+        }
+      }
+    }
+  }
+
+  /**
+   * Like announce(path, bytes, true).
+   */
+  public void announce(String path, byte[] bytes)
+  {
+    announce(path, bytes, true);
+  }
+
+  /**
+   * Announces the provided bytes at the given path.  Announcement means that it will create an ephemeral node
+   * and monitor it to make sure that it always exists until it is unannounced or this object is closed.
+   *
+   * @param path                  The path to announce at
+   * @param bytes                 The payload to announce
+   * @param removeParentIfCreated remove parent of "path" if we had created that parent
+   */
+  public void announce(String path, byte[] bytes, boolean removeParentIfCreated)
+  {
+    synchronized (toAnnounce) {
+      if (!started) {
+        toAnnounce.add(new Announceable(path, bytes, removeParentIfCreated));
+        return;
+      }
+    }
+
+    final String parentPath = ZKPathsUtils.getParentPath(path);
+    boolean buildParentPath = false;
+
+    byte[] value = announcedPaths.get(path);
+
+    if (value == null) {
+      try {
+        if (curator.checkExists().forPath(parentPath) == null) {
+          buildParentPath = true;
+        }
+      }
+      catch (Exception e) {
+        log.debug(e, "Problem checking if the parent existed, ignoring.");
+      }
+
+      // Synchronize to make sure that I only create a listener once.
+      synchronized (toAnnounce) {
+        if (!listeners.containsKey(path)) {
+          final NodeCache cache = new NodeCache(curator, path, true);
+          cache.getListenable().addListener(
+              () -> {
+                ChildData currentData = cache.getCurrentData();
+                if (currentData == null) {
+                  final byte[] value1 = announcedPaths.get(path);
+                  if (value1 != null) {
+                    log.info("Node[%s] dropped, reinstating.", path);
+                    createAnnouncement(path, value1);
+                  }
+                }
+              }
+          );
+
+          if (started) {
+            if (buildParentPath) {
+              createPath(parentPath, removeParentIfCreated);
+            }
+            startCache(cache);
+            listeners.put(path, cache);
+          }
+        }
+      }
+    }
+
+    boolean created = false;
+    synchronized (toAnnounce) {
+      if (started) {
+        byte[] oldBytes = announcedPaths.putIfAbsent(path, bytes);
```

Comment:
Please move the logic into `compute()` so when `IAE` is thrown the state is not changed.

Reply:
And if you make `started` volatile you can also drop the surrounding `synchronized (toAnnounce)`.

Reply:
The above suggestion is bad if we assume that `announce()` could be called concurrently with stop(). If this is the case, this should be explained in concurrent control flow documentation for this class and/or methods.

Reply:
> Please move the logic into compute() so when IAE is thrown the state is not changed.

@asdf2014 please do this